docs/development/testguide.rst

```diff
@@ -657,14 +657,28 @@ a real-world example::
       assert ("In merged column 'a' the 'units' attribute does not match (cm != m)"
               in str(warning_lines[0].message))
 
+`pytest`_ provides its own context manager
```

Comment:
This change needs to be in a separate PR as it is entirely unrelated to fixing an issue with the fast parser.  In a separate PR it can be reviewed by the right people.  

Personally I would reduce the first paragraph to an additional sentence in the note saying roughly "Do not use `pytest.warns`, always use the astropy version".  Unless there is a really good reason to use the pytest one, it is better to recommend one and only one way that we test warnings.

Reply:
This emerged indeed as a side product of adding warning checks to the tests here, and reviewing a load of `pytest.warns` tests for #9087 at the same time.
I only at that point noticed that the existing section practically did not cover `pytest.raises` at all, but just compared `astropy.tests.helper.catch_warnings` to the builtin `warnings.catch_warnings`. The case to prefer the former over `pytest.warns` does not seem nearly as clear cut. Anyway

1. There are plenty of existing `pytest.warns` in the tests right now, and generally with a legitimate use case, I'd say.

2. The actual warning about problems with `pytest` only seems to apply to the `recwarn` fixture (and that appears rather anecdotal and dated; from what I could still extract from #1174 the culprit might as well be [pytest.deprecated_call](https://docs.pytest.org/en/latest/warnings.html#ensuring-code-triggers-a-deprecation-warning), or perhaps the combination of both).

So the best recommendations I could extract from this situation were

1. Use `pytest.warns` wherever you expect one mandatory warning of known type (and possibly message).

2. Use `astropy.tests.helper.catch_warnings` to collect any number of warnings, not necessary exactly known ahead. Always use this over `warnings.catch_warnings`, `pytest.recwarn` or `with pytest.warns(None) as w`.

I guess this can use a bit more discussion on how to best phrase it, and of course ultimately it's a policy decision what to recommend, so I agree that this commit better goes into its own PR.